Assets/MixedRealityToolkit.SDK/Experimental/Features/Input/Handlers/PrimitiveManipulator.cs

```diff
@@ -0,0 +1,744 @@
+            foreach (var p in pointerIdToPointerMap.Values)
+            {
+                // Check pointer has a valid controller (e.g. gaze pointer doesn't)
+                if (p.pointer.Controller != null)
+                {
+                    numControllers++;
+                    sum += p.pointer.Controller.AngularVelocity;
+                }
+            }
+            return sum / Math.Max(1, numControllers);
+        }
+
+        private bool IsNearManipulation()
+        {
+            foreach (var item in pointerIdToPointerMap)
+            {
+                if (item.Value.IsNearPointer)
+                {
+                    return true;
+                }
+            }
+            return false;
+        }
+        #endregion Private Methods
+
+        #region Public Methods
+
+        /// <summary>
+        /// Releases the object that is currently manipulated
+        /// </summary>
+        public void ForceEndManipulation()
+        {
+            // end manipulation
+            if (isManipulationStarted)
+            {
+                HandleManipulationEnded(GetPointersGrabPoint(), GetPointersVelocity(), GetPointersAngularVelocity());
+            }
+            pointerIdToPointerMap.Clear();
+        }
+
+        /// <summary>
+        /// Gets the grab point for the given pointer id.
+        /// Only use if you know that your given pointer id corresponds to a pointer that has grabbed
+        /// this component.
+        /// </summary>
+        public Vector3 GetPointerGrabPoint(uint pointerId)
+        {
+            Assert.IsTrue(pointerIdToPointerMap.ContainsKey(pointerId));
+            return pointerIdToPointerMap[pointerId].GrabPoint;
+        }
+
+        #endregion Public Methods
+
+        #region Hand Event Handlers
+
+        /// <inheritdoc />
+        public void OnPointerDown(MixedRealityPointerEventData eventData)
+        {
+            if (!allowFarManipulation && eventData.Pointer as IMixedRealityNearPointer == null)
+            {
+                return;
+            }
+
+            // If we only allow one handed manipulations, check there is no hand interacting yet. 
+            if (manipulationType != HandMovementType.OneHanded || pointerIdToPointerMap.Count == 0)
+            {
+                uint id = eventData.Pointer.PointerId;
+                // Ignore poke pointer events
+                if (!eventData.used && !pointerIdToPointerMap.ContainsKey(id))
+                {
+                    // cache start ptr grab point
+                    pointerIdToPointerMap.Add(id, new PointerData(eventData.Pointer, eventData.Pointer.Result.Details.Point));
+
+                    // Call manipulation started handlers
+                    var handsPressedCount = pointerIdToPointerMap.Count;
+                    if (manipulationType.HasFlag(HandMovementType.TwoHanded) && handsPressedCount > 1)
+                    {
+                        if (!manipulationType.HasFlag(HandMovementType.OneHanded))
```

Comment:
>!manipulationType.HasFlag(HandMovementType.OneHanded)) [](start = 28, length = 54)

I think that this line makes some assumptions about ordering with the else if statement down below (i.e. that the one handed manipulation thing would have been engaged first - totally valid, though this is also somewhat of a state machine in itself - thoughts on just having:

```csharp
if (!isManipulationStarted)
{
  HandleManipulationStarted();
}
```